modules/services/syncthing.nix

```diff
@@ -5,52 +5,585 @@ with lib;
+        # should update that device/folder only.
+        lib.pipe s.conf [
+          # Quoting https://docs.syncthing.net/rest/config.html:
+          #
+          # > PUT takes an array and POST a single object. In both cases if a
+          # given folder/device already exists, it’s replaced, otherwise a new
+          # one is added.
+          #
+          # What's not documented, is that using PUT will remove objects that
+          # don't exist in the array given. That's why we use here `POST`, and
+          # only if s.override == true then we DELETE the relevant folders
+          # afterwards.
+          (map (new_cfg: ''
+            curl -d ${
+              lib.escapeShellArg (builtins.toJSON new_cfg)
+            } -X POST ${s.baseAddress}
+          ''))
+          (lib.concatStringsSep "\n")
+        ]
+        /* If we need to override devices/folders, we iterate all currently configured
+           IDs, via another `curl -X GET`, and we delete all IDs that are not part of
+           the Nix configured list of IDs
+        */
+        + lib.optionalString s.override ''
+          stale_${conf_type}_ids="$(curl -X GET ${s.baseAddress} | ${jq} \
+            --argjson new_ids ${
+              lib.escapeShellArg (builtins.toJSON s.new_conf_IDs)
+            } \
+            --raw-output \
+            '[.[].${s.GET_IdAttrName}] - $new_ids | .[]'
+          )"
+          for id in ''${stale_${conf_type}_ids}; do
+            curl -X DELETE ${s.baseAddress}/$id
+          done
+        ''))
+      builtins.attrValues
+      (lib.concatStringsSep "\n")
+    ]) +
+    /* Now we update the other settings defined in cleanedConfig which are not
+       "folders" or "devices".
+    */
+    (lib.pipe cleanedConfig [
+      builtins.attrNames
+      (lib.subtractLists [ "folders" "devices" ])
+      (map (subOption: ''
+        curl -X PUT -d ${
+          lib.escapeShellArg (builtins.toJSON cleanedConfig.${subOption})
+        } ${curlAddressArgs "/rest/config/${subOption}"}
+      ''))
+      (lib.concatStringsSep "\n")
+    ]) + lib.optionalString (cfg.passwordFile != null) ''
+      syncthing_password=$(${cat} ${cfg.passwordFile})
+      curl -X PATCH -d '{"password": "'$syncthing_password'"}' ${curlAddressArgs "/rest/config/gui"}
+    '' + ''
+      # restart Syncthing if required
+      if curl ${curlAddressArgs "/rest/config/restart-required"} |
+         ${jq} -e .requiresRestart > /dev/null; then
+          curl -X POST ${curlAddressArgs "/rest/system/restart"}
+      fi
+    '');
 
   defaultSyncthingArgs = [
-    "${pkgs.syncthing}/bin/syncthing"
+    "${syncthing}"
     "-no-browser"
     "-no-restart"
+    "-no-upgrade"
+    "-gui-address=${if isUnixGui then "unix://" else ""}${cfg.guiAddress}"
     "-logflags=0"
   ];
 
   syncthingArgs = defaultSyncthingArgs ++ cfg.extraOptions;
-
 in {
   meta.maintainers = [ maintainers.rycee ];
 
-  options = {
+  options = with types; {
```

Comment:
At the moment, this PR enables the declarative Syncthing configuration only on Linux, not on Darwin. The documentation of the module should mention this at some place. Or how is this handled in other Home Manager modules?

As a related side note: I used your (very nice!) PR as a base in [my own home-manger fork](https://github.com/pitkling/home-manager/tree/syncthing-declarative) to add support for Darwin. The first 5 additional commits are mostly some refactoring to make it possible to reuse some of the code. The [last commit](https://github.com/nix-community/home-manager/commit/e40308638db88656394adef26970cc667b00076e) adds two launchd agents akin to the two systemd services you define in your PR.

I assume it might be cleaner to make the Darwin support a separate PR? Anyway, I'm still testing those commits, although so far everything seems to work fine :).

Reply:
I definitely think it should be a separate PR, I don't have a Darwin device to test things on and I wouldn't want to push untested code ;)